### PR TITLE
Move pipeline implementation to a separate object

### PR DESCRIPTION
**Pull request by happz**
A pipeline can spawn a new, derived pipeline. Glue instance then takes
care of managing them, discovering all registered shared functions and
similar top-level stuff.

Primary motivation was to let modules read pipeline description while in
destroy phase, and let modules spawn pipelines. It proved to be worth
the effort to refactor related code, clearing the interface.

### CODE REVIEW

**Comment by happz** [comment]
Not fully tested yet, and tests need fixing and refactoring to accommodate new structure, but it is definitely open for a review.

**Comment by thrix** [comment]
awesome, I will try ot understand how to initialize a new pipeline, so I can try it out ...

**Comment by happz** [comment]
> awesome, I will try ot understand how to initialize a new pipeline, so I can try it out ...

It's simple, things didn't change that much:

```python
glue.run_pipeline(glue.Pipeline([glue.PipelineStep('rules-engine')]))
```

is the same as:

```python
glue.run_modules([glue.PipelineStep('rules-engine')])
```

or, since it's just a single module, one can use:

```python
glue.run_module('rules-engine')
```

Documentation and tests will arrive soon, hopefuly.

File: gluetool/glue.py [hunk 1]
```diff
@@ -52,6 +54,22 @@
     '{}/gluetool_modules'.format(sys.prefix)
 ]
 
+#
+# NOTE: pipelines and shared functions.
+#
+# Each `Glue` instance manages one or more pipelines. The main one, requested by the user, can spawn additional
+# "side-car" pipelines, and these can spawn their own children, and so on. Only a single pipeline is being processed
+# at the moment - when a new pipeline is spawned, its parent gives up its time for its kid, and is awoken when the
+# kid finishes. `Glue` keeps a stack of pipelines as thei appear and disappear. There is always a "current pipeline",
```
Comment (mkluson) [inline gluetool/glue.py]: typo: thei

**Comment by thrix** [comment]
Cool, will test this properly PipelineCallbackStep looks awesome

**Comment by thrix** [comment]
Works for me!

**Review by thrix — DISMISSED** [review]
I tested, Used the callback. created separate pipeline. Everything seems fine ...

File: gluetool/glue.py [hunk 2]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+    """
+    Pipeline of ``gluetool`` modules. Defined by its steps, takes care of registering their shared functions,
+    running modules and destroying the pipeline.
+
+    To simplify the workflow, 2 primitives are defined:
+
+    * :py:meth:`_safe_call` - calls a given callback, returns its return value. Any exception raised by the callback
+      is wrapped by :py:class:`Failure` instance and returned instead of what callback would return.
+    * :py:meth:`_for_each_module` - loop over given list of modules, calling given callback for each of the modules.
+      :py:meth:`_safe_call` is used for the call, makign sure we always have a return value and no exceptions.
```
Comment (mkluson) [inline gluetool/glue.py]: typo: makign

File: gluetool/glue.py [hunk 3]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+
+        # pylint: disable=broad-except
+        except Exception:
+            return Failure(module=self.current_module, exc_info=sys.exc_info())
+
+    def _for_each_module(self, modules, callback, *args, **kwargs):
+        # type: (Iterable[Module], Callable[..., Optional[Failure]], *Any, **Any) -> Optional[Failure]
+        """
+        For each module in a list, call a given function with module as its first argument. If the call
+        returns anythign but ``None``, the value is returned by this function as well, ending the loop.
```
Comment (mkluson) [inline gluetool/glue.py]: typo: anythign

File: gluetool/glue.py [hunk 4]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+            exceptions.
+        :returns: value returned by ``callback``, or ``None`` when loop finished.
+        """
+
+        for module in modules:
+            self.current_module = module
+
+            # Given that we're using `_safe_call`, we shouldn't encounter any exception - `_safe_call`
+            # would convert any exception into `Failure` instance. Callback also cannot return either
+            # `None` or a failure. Therefore, if we gat anything `True`-ish, we simply pass it to our
```
Comment (mkluson) [inline gluetool/glue.py]: typo: gat

File: gluetool/glue.py [hunk 5]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+            # caller since it must by a failure.
+            #
+            # Note: This is enforced by type checks, we have no other power over the callback and its
+            # return typ.e
```
Comment (mkluson) [inline gluetool/glue.py]: typo: typ.e

File: gluetool/glue.py [hunk 6]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+
+        return None
+
+    def _setup(self):
+        # type: () -> Optional[Failure]
+
+        # Make a copy of steps - while setting modules up, we won't have access to module index, so we cannot
+        # reach to `self.steps` for its arguments, but we can pop the first item of this list - it's always
+        # the "current" module, the one currently being set up.
+        steps = self.steps[:]
```
Comment (mkluson) [inline gluetool/glue.py:608]: what is the difference between this and implementation and `self.steps.copy()`?

File: gluetool/glue.py [hunk 7]
```diff
@@ -327,9 +361,423 @@ def serialize_to_json(self):
...
+            # Module's shared functions are registered after its `execute` method finishes. But we want to register
+            # module's shared function no matter whether its `execute` finished successfully, or crashed. We could
+            # wrap `execute` call with `try` and `finally``, and call module's `add_shared` there but should there
+            # be an exception under `try` (which can) **and** should there be another one under `finally` (oh, yes,
+            # it can!), the first exception would be lost, replaced by the later. And we cannot guarantee exception-less
+            # `add_shared`, it may have been replaced by module's developer.
+            #
+            # Therefore, we catch the (possible) exception raised by `execute`, save its info aside, register shared
+            # functions, and re-raise the exception. By the act of saving the exception info, we let our exception
+            # chaining to catch up with us - should `add_shared` raise any exception, it would correctly recognize
```
Comment (mkluson) [inline gluetool/glue.py:654]: typo: corrently

File: gluetool/glue.py [hunk 8]
```diff
@@ -1020,68 +1508,87 @@ def parse_args(self, args):
...
-        By default, this method does nothing. Reimplement as needed.
+        # A proxy for Glue's `shared`, exists to simplify modules.
+
+        return self.glue.shared(funcname, *args, **kwargs)
+
+    def overloaded_shared(self, funcname, *args, **kwargs):
+        # type: (str, *Any, **Any) -> Any
         """
+        Call a shared function overloaded by the one provided by this module. This way,
+        a module can give chance to other implementations of its action, e.g. to publish
```
Comment (mkluson) [inline gluetool/glue.py]: do not mention message bus here, it is to tied to our usecases

File: gluetool/glue.py (hunk 6 above)
Comment (ukulekek) [reply gluetool/glue.py:608]: They are equal. Nevertheless, `list.copy()` is not available in python 2.7

Comment (happz) [reply gluetool/glue.py:608]: Right. Both work the same way, I'm used to `[:]`.

File: gluetool/glue.py (hunk 7 above)
Comment (happz) [reply gluetool/glue.py:654]: This I cannot find, looks got to me.

File: gluetool/glue.py (hunk 8 above)
Comment (happz) [reply gluetool/glue.py]: I think it's a legitimate example, we're not the only project in the world sending messages to a message bus.

Comment (happz) [reply gluetool/glue.py]: @mkluson I tried to rewrite the wording slightly, let me know your opinion (or resolve the conversation).

Comment (happz) [reply gluetool/glue.py]: @mkluson so, how are we standing on the review of my update?

File: gluetool/tests/test_run_modules.py [hunk 9]
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
...
+
+    assert isinstance(ret, gluetool.Failure)
+    assert ret.module is None
+    assert ret.exception is mock_exc
+    assert isinstance(ret.exc_info, tuple)
+
+
+def test_for_each_module(pipeline):
+    modules = [
+        create_module(DummyModule, name=DummyModule.name)[1],
```
Comment (mkluson) [inline gluetool/tests/test_run_modules.py:112]: Can `module` fixture be used here?

File: gluetool/tests/test_run_modules.py [hunk 10]
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
...
+        broken_module
+    ]
+
+    ret = pipeline._for_each_module(modules, lambda mod: mod.execute())
+
+    assert isinstance(ret, gluetool.Failure)
+    assert ret.exception.message == 'bar'
+
+
+def test_pipeline_setup(pipeline, monkeypatch):
```
Comment (mkluson) [inline gluetool/tests/test_run_modules.py]: param `monkeypatch` is unused in this function

File: gluetool/tests/test_run_modules.py [hunk 11]
```diff
@@ -45,41 +52,226 @@ def fixture_glue():
...
-        glue.run_modules([gluetool.glue.PipelineStep('Broken module')])
+    _test_add_shared(glue, broken_pipeline, monkeypatch, BrokenModule)
+
+
+def test_add_shared_missing(pipeline, module):
+    with pytest.raises(gluetool.GlueError, match=r"No such shared function 'does_not_exist' of module 'Dummy module"):
+        pipeline.add_shared('does_not_exist', module)
+
+
+def test_has_shared(glue, pipeline, module):
```
Comment (mkluson) [inline gluetool/tests/test_run_modules.py]: param `module` is unused

File: gluetool/tests/test_run_modules.py (hunk 9 above)
Comment (happz) [reply gluetool/tests/test_run_modules.py:112]: No, it must be two different instances (`assert module is expected_module`). I could use `module` for one instance and create the other, it seemed to be simpler to create both of them.

File: gluetool/tests/test_run_modules.py (hunk 10 above)
Comment (happz) [reply gluetool/tests/test_run_modules.py]: Fixed.

File: gluetool/tests/test_run_modules.py (hunk 11 above)
Comment (happz) [reply gluetool/tests/test_run_modules.py]: Fixed.

**Review by mkluson — APPROVED** [review]

**Review by thrix — APPROVED** [review]
LGTM, tested, using in almost production :D

**Comment by thrix** [comment]
@happz  I believe we can merge this without pull approve .. this requerest have been reviewed as needed ...

**Pull request merged**